### PR TITLE
document serving's revision garbage collection

### PR DESCRIPTION
**Pull request by dprotaso**

### CODE REVIEW

File: config/nav.yml [hunk 1]
```diff
@@ -51,6 +51,7 @@ nav:
         - Uninstalling Knative: admin/install/uninstall.md
         # Serving config
         - Knative Serving configuration:
+          - Revision Garbage Collection: admin/serving/revision-gc.md
```
Comment (abrennan89) [inline config/nav.yml:54]:
```suggestion
          - Revision garbage collection: admin/serving/revision-gc.md
```
titles should be sentence case unless it's a proper name of an object or something

File: config/nav.yml [hunk 2]
```diff
@@ -119,6 +120,7 @@ nav:
         - Load balancing:
           - Overview: serving/load-balancing/README.md
           - Configuring target burst capacity: serving/load-balancing/target-burst-capacity.md
+      - Revision Garbage Collection: serving/revision-gc.md
```
Comment (abrennan89) [inline config/nav.yml:123]:
```suggestion
      - Revision garbage collection: serving/revision-gc.md
```

Comment (abrennan89) [reply config/nav.yml:123]: why add this here instead of the developer or admin guide?

File: docs/admin/serving/revision-gc.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Revision Garbage Collection
+
+The `config-gc` ConfigMap contains settings that determine how to clean up in-active revisions. This ConfigMap is located in the `knative-serving` namespace.
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:3]:
```suggestion
The `config-gc` ConfigMap contains settings that determine how inactive revisions are cleaned up. This ConfigMap is located in the `knative-serving` namespace.
```

File: docs/admin/serving/revision-gc.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# Revision Garbage Collection
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:1]:
```suggestion
# Revision garbage collection
```

File: docs/admin/serving/revision-gc.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+The `config-gc` ConfigMap contains settings that determine how to clean up in-active revisions. This ConfigMap is located in the `knative-serving` namespace.
+
+
+## Cluster-wide Configuration
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:6]:
```suggestion
## Cluster-wide configuration
```

File: docs/admin/serving/revision-gc.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+retain-since-last-active-time: "disabled"
+max-non-active-revisions: "10"
+```
+
+#### Disable GC
+```
+retain-since-create-time: "disabled"
+retain-since-last-active-time: "disabled"
+max-non-active-revisions: "disabled"
+```
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:46]: can you show some context for these, i.e. where in the configmap would they appear? I feel like these might be a bit too abstract as examples

File: docs/admin/serving/revision-gc.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+#### Complex Example
+
+This example config wll keep recently deployed or active revisions,
+always maintain the last two in case of rollback, and prevent burst
+activity from exploding the count of old revisions.
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:52]:
```suggestion
The following example configuration keeps recently deployed or active revisions,
always maintains the last two revisions in case of rollback, and prevents burst
activity from exploding the count of old revisions:
```

File: docs/admin/serving/revision-gc.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+retain-since-create-time: "48h"
+retain-since-last-active-time: "15h"
+min-non-active-revisions: "2"
+max-non-active-revisions: "1000"
+```
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:59]: Please show more context as per above comment - this doesn't show where in the configmap these would be added etc

File: docs/admin/serving/revision-gc.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+## Per Revision Options
+
+Users can alwaysretain a revision by adding the annotation `serving.knative.dev/no-gc: "true"`
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:63]:
```suggestion
Users can configure their settings so that at least one revision is always retained, by adding the `serving.knative.dev/no-gc: "true"` annotation to a Service:
```

File: docs/admin/serving/revision-gc.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+activity from exploding the count of old revisions.
+
+```
+retain-since-create-time: "48h"
+retain-since-last-active-time: "15h"
+min-non-active-revisions: "2"
+max-non-active-revisions: "1000"
+```
+
+## Per Revision Options
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:61]:
```suggestion
## Per-revision options
```

File: docs/admin/serving/revision-gc.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+#### Disable GC
+```
+retain-since-create-time: "disabled"
+retain-since-last-active-time: "disabled"
+max-non-active-revisions: "disabled"
+```
+
+#### Complex Example
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:48]:
```suggestion
#### Complex example
```

File: docs/serving/revision-gc.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# Revision Garbage Collection
```
Comment (abrennan89) [inline docs/serving/revision-gc.md:1]:
```suggestion
# Revision garbage collection
```

File: docs/serving/revision-gc.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+Knative will automatically clean up inactive revisions as configured by the operator. For more information, see the [operator settings](../admin/serving/revision-gc.md).
```
Comment (abrennan89) [inline docs/serving/revision-gc.md:3]:
```suggestion
Knative automatically cleans up inactive revisions as configured by the Operator. For more information, see the [Operator settings](../admin/serving/revision-gc.md).
```

File: docs/serving/revision-gc.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+To always retain a revision add the annotation `serving.knative.dev/no-gc: "true"`
```
Comment (abrennan89) [inline docs/serving/revision-gc.md:5]:
```suggestion
You can configure a Service so that at least one revision is always retained, by adding the `serving.knative.dev/no-gc: "true"` annotation:
```

File: docs/serving/revision-gc.md (hunk 12 above)
Comment (abrennan89) [reply docs/serving/revision-gc.md:1]: I'm not sure there's a need to duplicate this information here - IMO it would be enough to have the admin guide configmap based docs, and then have a section in the developer guide with the per-revision docs, rather than just duplicating info in the Knative Serving section that we've been trying to move into dev vs admin docs.

I won't block the PR on this though since the structure is already a mess, just something to think about.

**Review by abrennan89 — COMMENTED** [review]
Some suggestions to bring this in line with the doc guidelines and fix some typos etc.

File: docs/serving/revision-gc.md (hunk 12 above)
Comment (dprotaso) [reply docs/serving/revision-gc.md]: Yeah - I'm mixed on it - but made the separate section since the per-revision option may be available to end-users

File: docs/serving/revision-gc.md (hunk 14 above)
Comment (dprotaso) [reply docs/serving/revision-gc.md]: Updated the example to be more clear where the annotation is supposed to end up

**Comment by julz** [comment]
Think this “Fixes #3461“, thanks @dprotaso!

File: docs/admin/serving/revision-gc.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+## Cluster-wide configuration
+
+The following properties allow you to configure revision garbage collection:
+
+-|-
+name|description
+`retain-since-create-time`| Duration since creation before considering a revision for GC or "disabled"
+`retain-since-last-active-time`| Duration since active before considering a revision for GC or "disabled"
+`min-non-active-revisions`| Minimum number of non-active revisions to retain.
+`max-non-active-revisions`| Maximum number of non-active revisions to retain or "disabled" to disable any maximum limit.
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:15]: Presumably this should be a table but I think the formatting is wrong since it's not rendering in the preview?
https://deploy-preview-4266--knative.netlify.app/development/admin/serving/revision-gc/#cluster-wide-configuration

File: docs/admin/serving/revision-gc.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+Revisions are retained if they fall into any one of the following categories
+- Is active and is being reference by a route
+- Created within `retain-since-create-time`
+- Last referenced by a route within `retain-since-last-active-time`
+- There are fewer than `min-non-active-revisions`
+
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:22]:
```suggestion
Revisions are retained if they fall into any one of the following categories:

- Is active and is being reference by a route
- Created within `retain-since-create-time`
- Last referenced by a route within `retain-since-last-active-time`
- There are fewer than `min-non-active-revisions`
```

Fix line spacing so this renders properly in the preview

File: docs/admin/serving/revision-gc.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+metadata:
+  name: config-gc
+  namespace: knative-serving
+data:
+  retain-since-create-time: "disabled"
+  retain-since-last-active-time: "disabled"
+  max-non-active-revisions: "10"
+```
+
+#### Disable GC
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:52]:
```suggestion
#### Disable garbage collection
```
I'd spell this out tbh

File: docs/admin/serving/revision-gc.md [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+  retain-since-create-time: "disabled"
+  retain-since-last-active-time: "disabled"
+  max-non-active-revisions: "disabled"
+```
+
+#### Complex example
+
+Th following example configuration keeps recently deployed or active revisions,
+always maintains the last two revisions in case of rollback, and prevents burst
+activity from exploding the count of old revisions:
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:69]:
```suggestion
The following example configuration keeps recently deployed or active revisions,
always maintains the last two revisions in case of rollback, and prevents burst
activity from exploding the count of old revisions:
```
Minor typo

File: docs/admin/serving/revision-gc.md [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+data:
+  retain-since-create-time: "48h"
+  retain-since-last-active-time: "15h"
+  min-non-active-revisions: "2"
+  max-non-active-revisions: "1000"
+```
+
+## Per-revision options
+
+You can always retain a revision by adding the `serving.knative.dev/no-gc: "true"` annotation
```
Comment (abrennan89) [inline docs/admin/serving/revision-gc.md:86]:
```suggestion
You can configure a revision so that it is never garbage collected by adding the `serving.knative.dev/no-gc: "true"` annotation:
```
I'd reword this to make it clearer - the always retain bit just reads weird to me. Might be just me though.

File: docs/serving/revision-gc.md [hunk 20]
```diff
@@ -0,0 +1,15 @@
+# Revision garbage collection
+
+Knative automatically cleans up inactive revisions as configured by the Operator. For more information, see the [Operator settings](../admin/serving/revision-gc.md).
+
+You can always retain a revision by adding the `serving.knative.dev/no-gc: "true"` annotation
```
Comment (abrennan89) [inline docs/serving/revision-gc.md:5]:
```suggestion
You can configure a revision so that it is never garbage collected by adding the `serving.knative.dev/no-gc: "true"` annotation:
```
I'd reword this to make it clearer - the always retain bit just reads weird to me. Might be just me though.

**Review by abrennan89 — COMMENTED** [review]
Some final comments, mostly fixing formatting that's broken in the preview

File: docs/admin/serving/revision-gc.md (hunk 15 above)
Comment (dprotaso) [reply docs/admin/serving/revision-gc.md:15]: good catch - let me fix that

**Review by abrennan89 — APPROVED** [review]
/lgtm

Thanks @dprotaso !
